File: src/platform/domain/performance.py
"""Domain models for Strategy Performance Analytics and Risk Intelligence.

Immutable domain dataclasses representing strategy performance summary, drawdown profiles,
and risk analytics profiles.
"""

from dataclasses import dataclass, field
from typing import Any, Dict, Optional


@dataclass(frozen=True)
class DrawdownProfile:
    """Drawdown profile for strategy risk assessment."""

    max_drawdown_pct: float = 0.0
    recovery_factor: float = 0.0
    risk_classification: str = "LOW"

    def to_dict(self) -> Dict[str, Any]:
        return {
            "max_drawdown_pct": self.max_drawdown_pct,
            "recovery_factor": self.recovery_factor,
            "risk_classification": self.risk_classification,
        }


@dataclass(frozen=True)
class RiskAnalyticsProfile:
    """Risk analytics profile evaluating trade setup compliance and limits."""

    reward_to_risk_expectancy: float = 0.0
    max_drawdown: float = 0.0
    win_loss_ratio: float = 0.0
    risk_level: str = "medium"
    position_risk_pass: bool = True
    drawdown_limit_pass: bool = True
    risk_assessment_status: str = "compliant"

    def to_dict(self) -> Dict[str, Any]:
        return {
            "reward_to_risk_expectancy": self.reward_to_risk_expectancy,
            "max_drawdown": self.max_drawdown,
            "win_loss_ratio": self.win_loss_ratio,
            "risk_level": self.risk_level,
            "position_risk_pass": self.position_risk_pass,
            "drawdown_limit_pass": self.drawdown_limit_pass,
            "risk_assessment_status": self.risk_assessment_status,
        }


@dataclass(frozen=True)
class StrategyPerformanceSummary:
    """Aggregated strategy performance summary domain model."""

    strategy_name: str
    symbol: str = "XAUUSD"
    timeframe: str = "1h"
    status: str = "active"
    total_trades: int = 0
    win_rate: float = 0.0
    profit_factor: float = 0.0
    max_drawdown: float = 0.0
    net_profit: float = 0.0
    sharpe_ratio_estimate: Optional[float] = None
    drawdown_profile: Optional[DrawdownProfile] = None
    risk_profile: Optional[RiskAnalyticsProfile] = None
    walk_forward_efficiency: Optional[float] = None
    walk_forward_consistency: Optional[float] = None
    detail: str = ""
    metadata: Dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "strategy_name": self.strategy_name,
            "symbol": self.symbol,
            "timeframe": self.timeframe,
            "status": self.status,
            "total_trades": self.total_trades,
            "win_rate": self.win_rate,
            "profit_factor": self.profit_factor,
            "max_drawdown": self.max_drawdown,
            "net_profit": self.net_profit,
            "sharpe_ratio_estimate": self.sharpe_ratio_estimate,
            "drawdown_profile": self.drawdown_profile.to_dict() if self.drawdown_profile else None,
            "risk_profile": self.risk_profile.to_dict() if self.risk_profile else None,
            "walk_forward_efficiency": self.walk_forward_efficiency,
            "walk_forward_consistency": self.walk_forward_consistency,
            "detail": self.detail,
            "metadata": dict(self.metadata) if self.metadata else {},
        }

File: web/src/architecture/performance.ts
import type { HostSnapshot } from "./hostView";

export interface DrawdownProfileData {
  max_drawdown_pct: number;
  recovery_factor: number;
  risk_classification: "LOW" | "MEDIUM" | "HIGH" | "CRITICAL" | string;
}

export interface RiskAnalyticsProfileData {
  reward_to_risk_expectancy: number;
  max_drawdown: number;
  win_loss_ratio: number;
  risk_level: string;
  position_risk_pass: boolean;
  drawdown_limit_pass: boolean;
  risk_assessment_status: "compliant" | "degraded" | "breached" | string;
}

export interface StrategyPerformanceSummaryData {
  strategy_name: string;
  symbol: string;
  timeframe: string;
  status: string;
  total_trades: number;
  win_rate: number;
  profit_factor: number;
  max_drawdown: number;
  net_profit: number;
  sharpe_ratio_estimate?: number | null;
  drawdown_profile?: DrawdownProfileData | null;
  risk_profile?: RiskAnalyticsProfileData | null;
  walk_forward_efficiency?: number | null;
  walk_forward_consistency?: number | null;
  detail?: string | null;
}

export interface PerformanceAnalyticsState {
  status: "active" | "connected" | "disconnected" | "unavailable" | "empty" | "unauthorized" | "failed" | "loading";
  message: string;
  summary: StrategyPerformanceSummaryData | null;
}

export function extractPerformanceAnalytics(snapshot: HostSnapshot): PerformanceAnalyticsState {
  if (!snapshot.project1.connected) {
    return {
      status: "disconnected",
      message: "Project 1 is disconnected. Connect Project 1 to access strategy performance analytics.",
      summary: null,
    };
  }

  const perf = snapshot.performance as {
    status?: string;
    message?: string;
    data?: StrategyPerformanceSummaryData;
  };

  if (!perf) {
    return {
      status: "unavailable",
      message: "Performance analytics are unavailable until Project 1 outputs are evaluated.",
      summary: null,
    };
  }

  const statusStr = (perf.status || "unavailable").toLowerCase() as PerformanceAnalyticsState["status"];

  if (statusStr === "unauthorized") {
    return {
      status: "unauthorized",
      message: perf.message || "Access denied: performance analytics restricted.",
      summary: null,
    };
  }

  if (statusStr === "empty") {
    return {
      status: "empty",
      message: perf.message || "No performance data available for this market context.",
      summary: null,
    };
  }

  if (statusStr === "failed") {
    return {
      status: "failed",
      message: perf.message || "Performance analytics calculation encountered an exception.",
      summary: null,
    };
  }

  if (!perf.data) {
    return {
      status: "unavailable",
      message: perf.message || "Performance analytics are unavailable until Project 1 outputs are evaluated.",
      summary: null,
    };
  }

  return {
    status: statusStr === "active" ? "active" : "connected",
    message: perf.message || "Authentic strategy performance analytics evaluated.",
    summary: perf.data,
  };
}

/* Authenticated REST API Client Helpers */

export async function fetchPerformanceSummary(
  token: string | null,
  strategyName: string = "Project 1 Strategy",
  symbol: string = "XAUUSD",
  timeframe: string = "1h"
): Promise<{ success: boolean; performance?: StrategyPerformanceSummaryData; error?: string }> {
  try {
    const headers: Record<string, string> = { Accept: "application/json" };
    if (token) headers["Authorization"] = `Bearer ${token}`;

    const url = `/api/v1/performance/summary?strategy_name=${encodeURIComponent(strategyName)}&symbol=${encodeURIComponent(symbol)}&timeframe=${encodeURIComponent(timeframe)}`;
    const res = await fetch(url, { method: "GET", headers });
    const data = await res.json();

    if (res.ok && data.success) {
      return { success: true, performance: data.performance?.data || data.performance };
    }
    return { success: false, error: data.why || data.message || "Failed to fetch performance summary" };
  } catch (err: any) {
    return { success: false, error: err.message || "Network error fetching performance summary" };
  }
}

export async function fetchRiskProfile(
  token: string | null,
  strategyName: string = "Project 1 Strategy",
  symbol: string = "XAUUSD",
  timeframe: string = "1h"
): Promise<{ success: boolean; risk?: RiskAnalyticsProfileData; drawdown?: DrawdownProfileData; error?: string }> {
  try {
    const headers: Record<string, string> = { Accept: "application/json" };
    if (token) headers["Authorization"] = `Bearer ${token}`;

    const url = `/api/v1/performance/risk?strategy_name=${encodeURIComponent(strategyName)}&symbol=${encodeURIComponent(symbol)}&timeframe=${encodeURIComponent(timeframe)}`;
    const res = await fetch(url, { method: "GET", headers });
    const data = await res.json();

    if (res.ok && data.success) {
      return { success: true, risk: data.risk, drawdown: data.drawdown };
    }
    return { success: false, error: data.why || data.message || "Failed to fetch risk profile" };
  } catch (err: any) {
    return { success: false, error: err.message || "Network error fetching risk profile" };
  }
}
